File: src/atcoder/abc314/swift/abc314_e.swift
func get2() -> (Int, Int) {
  let line = readLine()!.split(separator: " ").map { Int($0)! }
  return (line[0], line[1])
}

let (n, m) = get2()
var memo = [Double?](repeating: nil, count: m)

let roulettes = (0..<n).map { _ -> (Int, Int, Int, [Int]) in
  let line = readLine()!.split(separator: " ").map { Int($0)! }
  let c = line[0]
  let p = line[1]
  let s = Array(line[2...]).filter { v -> Bool in return v != 0 }
  let z = p - s.count
  return (c, p, z, s)
}

func dfs(memo : inout [Double?], roulettes : [(Int,Int,Int,[Int])], x : Int) -> Double {
  if x >= m {
    return 0.0
  }

  if let ans = memo[x] {
    return ans
  }

  var ans = 1e100

  for (c, p, z, s) in roulettes {
    var cnt = 0.0
    for y in s {
      cnt += dfs(memo: &memo, roulettes: roulettes, x: x + y)
    }
    cnt = (cnt + Double(p * c)) / Double(p - z)
    if ans > cnt {
      ans = cnt
    }
  }
  memo[x] = ans
  return ans
}

let ans = dfs(memo: &memo, roulettes: roulettes, x: 0)
print(ans)

File: src/atcoder/abc314/scala/abc314_e.scala
import io.StdIn.readLine
import scala.collection.mutable.ArrayBuffer

val buf = java.util.StringTokenizer(io.Source.stdin.mkString)
def sc = buf.nextToken

case class Roulettes(c:Int, p:Int, s:ArrayBuffer[Int], z:Int)

object Main extends App:
  val n, m = sc.toInt
  var memo = ArrayBuffer.fill(m) { None: Option[Double] }
  val roulettes = ArrayBuffer.fill(n) {
    val c = sc.toInt
    val p = sc.toInt
    val s = ArrayBuffer.fill(p) { sc.toInt }.filter(_ != 0)
    val z = p - s.size
    Roulettes(c, p, s, z)
  }

  def dfs(x:Int): Double =
    if x >= m then
      return 0.0
    
    memo(x) match
      case Some(ans) => ans
      case None =>
        var ans = 1e100

        for roulette <- roulettes do
          var cnt = 0.0

          val c = roulette.c
          val p = roulette.p
          val s = roulette.s
          val z = roulette.z

          for y <- s do
            cnt += dfs(x + y)

          cnt = (cnt + p * c) / (p - z)
          if ans > cnt then
            ans = cnt
        
        memo(x) = Some(ans)
        ans

  val ans = dfs(0)
  print(ans)
